src/mailmap.c

```diff
@@ -0,0 +1,335 @@
+/*
+ * Copyright (C) the libgit2 contributors. All rights reserved.
+ *
+ * This file is part of libgit2, distributed under the GNU GPL v2 with
+ * a Linking Exception. For full terms see the included COPYING file.
+ */
+
+#include "git2/mailmap.h"
+
+#include "blob.h"
+#include "commit.h"
+#include "parse.h"
+#include "git2/common.h"
+#include "git2/repository.h"
+#include "git2/revparse.h"
+#include "git2/sys/commit.h"
+
+#define MAILMAP_FILE ".mailmap"
+
+struct git_mailmap {
+	git_vector entries;
+};
+
+/* Check if we're at the end of line, w/ comments */
+static bool is_eol(git_parse_ctx *ctx)
+{
+	char c;
+	return git_parse_peek(&c, ctx, GIT_PARSE_PEEK_SKIP_WHITESPACE) < 0 || c == '#';
+}
+
+static int advance_until(
+	const char **start, size_t *len, git_parse_ctx *ctx, char needle)
+{
+	*start = ctx->line;
+	while (ctx->line_len > 0 && *ctx->line != '#' && *ctx->line != needle)
+		git_parse_advance_chars(ctx, 1);
+
+	if (ctx->line_len == 0 || *ctx->line == '#')
+		return -1; /* end of line */
+
+	*len = ctx->line - *start;
+	git_parse_advance_chars(ctx, 1); /* advance past needle */
+	return 0;
+}
+
+/* Parse a single entry from a mailmap file.
+ *
+ * The output git_bufs will be non-owning, and should be copied before being
+ * persisted.
+ */
+static int parse_mailmap_entry(
+	git_buf *real_name, git_buf *real_email,
+	git_buf *replace_name, git_buf *replace_email,
+	git_parse_ctx *ctx)
+{
+	const char *start;
+	size_t len;
+
+	git_buf_clear(real_name);
+	git_buf_clear(real_email);
+	git_buf_clear(replace_name);
+	git_buf_clear(replace_email);
+
+	/* Parse the real name */
+	git_parse_advance_ws(ctx);
+	if (advance_until(&start, &len, ctx, '<') < 0)
+		return -1;
```

Comment:
```
if (...) {
    giterr_set(GITERR_MAILMAP, "could not parse real name");
    return -1;
}
```

You'd have to define `GITERR_MAILMAP`, but I think it can be helpful to detail what's wrong here

Reply:
We always ignore these parser errors & just skip the line (hence the potential desire for warnings). Should we still set giterr if we won't produce a visible error code?

Reply:
I realized that a bit later and thought to have deleted that comment. Existing code is fine as-is.